### ISSUE TITLE
Upgrading Hilt 2.42 -> 2.44 adds a new deprecation warning

### ISSUE THREAD

**Issue by liutikas** [google/dagger#3601]
Upgrading Hilt 2.42 -> 2.44 adds a new deprecation warning.

When we upgraded androidx from Hilt 2.42 to 2.44 (https://android-review.googlesource.com/c/platform/frameworks/support/+/2242564) we started getting a new compile warning:
```
  > Task :hilt:hilt-navigation-fragment:hiltJavaCompileDebugAndroidTest
  $OUT_DIR/androidx/hilt/hilt-navigation-fragment/build/generated/hilt/component_sources/debugAndroidTest/androidx/hilt/navigation/fragment/HiltNavGraphViewModelLazyTest_TestComponentDataSupplier.java:16: warning: [deprecation] applicationContextModule(ApplicationContextModule) in Builder has been deprecated
          .applicationContextModule(
  1 warning
```

Test class https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:hilt/hilt-navigation-fragment/src/androidTest/java/androidx/hilt/navigation/fragment/HiltNavGraphViewModelLazyTest.kt;l=62

**Comment by Chang-Eric** [comment]
Hm, I suspect this might have to do with if something change such that nothing ever uses the `@ApplicationContext Context` binding. Could you try adding a usage of that binding in the test and see if the deprecation warning goes away?

**Comment by danysantiago** [comment]
For Dagger its fine if it adds `@Deprecated` to builders methods when the binding is not used since one can control the builder invocations, but if using Hilt one has no control when building the component so I think Hilt generated code should suppress the warning.

This is the generated file:
```
public final class HiltNavGraphViewModelLazyTest_TestComponentDataSupplier extends TestComponentDataSupplier {
  protected TestComponentData get() {
    return new TestComponentData(false, testInstance -> injectInternal((HiltNavGraphViewModelLazyTest) testInstance), Collections.emptySet(), Collections.emptySet(), (modules, testInstance, autoAddModuleEnabled) -> DaggerDefault_HiltComponents_SingletonC.builder()
         // ---> Calling a method that is marked as @Deprecated <---
        .applicationContextModule(
            new ApplicationContextModule(Contexts.getApplication(ApplicationProvider.getApplicationContext())))
        .build());
  }

  @SuppressWarnings("unchecked")
  private static void injectInternal(HiltNavGraphViewModelLazyTest testInstance) {
    ((HiltNavGraphViewModelLazyTest_GeneratedInjector) ((GeneratedComponentManager) Contexts.getApplication(ApplicationProvider.getApplicationContext())).generatedComponent()).injectTest(testInstance);
  }
}
```

**Comment by Chang-Eric** [comment]
Yea, I am looking to confirm that this is the actual cause of the problem before looking how to suppress this in Hilt.

**Comment by Chang-Eric** [comment]
Thanks, I think the right thing to do here then is to make Dagger not add that deprecation for bindings based on pruning, but just on whether the module itself will ever need an instance (i.e. if it only has static/abstract members).

**Comment by TWiStErRob** [comment]
I'm integrating Hilt for the first time to an existing project, and being greeted with this warning in code I didn't write is less than ideal.

Thanks @kuanyingchou for the workaround! For future users finding this, you can add a usage in the App too, no need for a ViewModel:
```java
@HiltAndroidApp
public class App extends Application {
	// TODO Delete when https://github.com/google/dagger/issues/3601 is resolved.
	@Inject @ApplicationContext Context context;
```

**Comment by ffimnsr** [comment]
This still happens today with the newly setup hilt android project. The current version is 2.51. Thanks for the current workaround @kuanyingchou and @TWiStErRob. Waiting for fixes on dagger#3601.

**Comment by JyotimoyKashyap** [comment]
Thanks for the workaround @TWiStErRob

**Comment by Inventrohyder** [comment]
Thanks for the workaround too @TWiStErRob @kuanyingchou @PatryMis @sohyun127

Still getting the issue with version `2.57.1`

Still waiting on the fix

**Comment by Erdou** [comment]
Hey guys, in which version of Dagger is this fixed? It's not tagged here. Or how do I know that from the repo?

**Comment by bcorso** [comment]
**tl;dr;** there hasn't been a release since this was fixed, so you would have to use the [`HEAD-SNAPSHOT`](https://dagger.dev/dev-guide/versions) artifacts to get this change.

------------

However, in general, to see which release a commit is in you can click the commit that closed the issue (see above):

<img width="498" height="54" alt="Image" src="https://github.com/user-attachments/assets/2136353b-20ed-4d6d-b3e0-ede5f4e2471d" />

Then you can see which tags the commit belongs to:

<img width="2223" height="608" alt="Image" src="https://github.com/user-attachments/assets/b5abfed6-f9bf-4b83-b30b-19906b43237d" />

In this case the commit doesn't belong to any tags because there hasn't been a release since this was fixed. Once we do a release the tag will show up, e.g.

<img width="431" height="53" alt="Image" src="https://github.com/user-attachments/assets/bb64f075-e2bd-4b70-b9f3-e7c9fa740c83" />
